### ISSUE TITLE
External keyboard (Bluetooth and USB) not recognised by browser window for Quest 3

### ISSUE THREAD

**Issue by kierannolan** [Igalia/wolvic#1141]
## Configuration

Wolvic version: 1.5.1
Wolvic build ID:  Meta store version

Hardware: Meta Quest 3

## Steps to Reproduce

1. Can type into the URL bar, but the browser window itself isn't responding to keyboard interactions

## Current Behavior

No response from the browser window to key inputs

## Expected Behavior

Feedback from keyboard inputs

```
```

</details>

**Comment by svillar** [comment]
Is it working fine in other apps or just Wolvic?

**Comment by kierannolan** [comment]
It works in the Meta browser but not Wolvic.

Here's a link to a sample scene for test purposes:
[](https://aestheticsvr.neocities.org/kbwebxr/index.html)

If you can implement this please it would be a huge help.

Also I'm including the full code below if you want to test it from a different webhost:


<html>
  
  <head>
    <script src="https://aframe.io/releases/1.5.0/aframe.min.js">
    </script>
  </head>
  
  <body>
    <a-scene xr-mode-ui="XRMode: xr">
      <a-box position="-1 0.5 -3" rotation="0 45 0" color="#4CC3D9">
      </a-box>
      <a-plane id="text-plane" position="0 1.5 -3" rotation="0 0 0" color="#FFF"
      width="2" height="1">
        <a-text id="display-text" value="Press a key" align="center" color="#000">
        </a-text>
      </a-plane>
    </a-scene>
    <script>
      //keytest code
      document.addEventListener('keydown',
      function(event) {
        var textElement = document.querySelector('#display-text');
        textElement.setAttribute('value', `Key Pressed: $ {
          event.key
        }`);
      });
    </script>
  </body>

</html>

**Comment by pt1997** [comment]
The keyboard inputs are recognized when you click on a text filed on a side, then you can also use "Shift"+"Arrow-keys" to select text. In this situation, the Volume control buttons on the headset don't work anymore.

I think this might be connected. It appears as if all input either goes to the page or to the application. If this is the case, this would need to be changed to make keyboard shortcuts like ctrl+tab, ctrl+f, F6, ... useable.

In addition, the following would be very useful to find the keyboard and the keys.
https://developer.oculus.com/documentation/native/android/mobile-physical-tracked-keyboard/

**Comment by boutell** [comment]
I am also experiencing this issue. As @pt1997 points out, interacting with a text field allows use of the keyboard, but this does not account for applications like vscode.dev / code server (web-hosted Visual Studio Code) that rely on access to keyboard events any time the browser window has the focus. For me that means Wolvic is not currently an option as a productivity browser, which is a pity because it is more flexible than the Meta browser in other ways.

I'll also note that even when working with a text field you just selected, putting down your hand controllers usually results in a loss of keyboard focus, either right away or after a few seconds.

Wolvic also doesn't appear to support a mouse pointer, which is supported both on the main Quest desktop and in the Quest browser.

**Comment by boutell** [comment]
Of course I realize this is an open source project. Just a thought, is there a way to chip in toward particular features?

**Comment by pt1997** [comment]
@boutell I don't know about this, but you can donate here https://opencollective.com/wolvic

**Comment by boutell** [comment]
Thanks. I made a small contribution, just because I like the project.

**Comment by jennie-bell** [comment]
I'm not sure how applicable it is now, or whether anything has changed, but I couldn't help but notice something lately revolving this issue:

If you're using a bluetooth keyboard, select a text field, and attempt to type, chances are that it's not presenting you with the text you typed because when you put your controller(s) down, your cursor hovered--or is hovering--over another window, and Wolvic unfocused that text field and will only refocus it when you hover over the original window again, regardless of whether you actually focused any other windows.

Try it -- Select a text field and, with one hand, press keys on your keyboard, and with your other, hover your cursor over the focused and unfocused windows. It only registers the characters you typed while your cursor was on the window of focus.

**Comment by svillar** [comment]
> I'm not sure how applicable it is now, or whether anything has changed, but I couldn't help but notice something lately revolving this issue:
> 
> If you're using a bluetooth keyboard, select a text field, and attempt to type, chances are that it's not presenting you with the text you typed because when you put your controller(s) down, your cursor hovered--or is hovering--over another window, and Wolvic unfocused that text field and will only refocus it when you hover over the original window again, regardless of whether you actually focused any other windows.
> 
> Try it -- Select a text field and, with one hand, press keys on your keyboard, and with your other, hover your cursor over the focused and unfocused windows. It only registers the characters you typed while your cursor was on the window of focus.

Yep, we definitely need to improve that experience, we should be able to turn off the controllers' aim while typing. However this is not happening soon because Meta is going to remove the OpenXR extensions that support this use case. Actually it's in the process of replacing them by others, so we need to wait till things settle down.
